Feature/docs

maslinych opened the pull request:
Руководство по i-bem.js. Готово к мерджу.

maslinych committed:
start i-bem.js reference

maslinych committed:
WIP on i-bem.js reference

maslinych committed:
WIP on i-bem.js reference

maslinych committed:
WIP i-bem.js reference

    section on block states

maslinych committed:
i-bem.js reference: events section

maslinych committed:
i-bem.js reference: modifiers

maslinych committed:
i-bem.js reference: IBC section

maslinych committed:
i-bem.js docs: reference section

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+
+ * jQuery. При использовании bem-core отдельная установка jQuery не требуется.
+ * Модульная система [ymaps/modules](https://github.com/ymaps/modules). 
+ * ymaps-модули `inherit`, `next-tick`, `vow` из библиотеки bem-core.
```
Это не ymaps-модули

maslinych replied:
А как их назвать, просто «модули из библиотеки bem-core»?

dfilatov replied:
А зачем тут вообще описывать зависимости от этих модулей?

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+
+
+В результате при инциализации блоков будет создан один JS-объект, содержащий ссылки на оба DOM-узла. 
+_!!уточнить, верно ли!!_
```
верно

maslinych replied:
А как хранятся эти ссылки? Массив  в поле domElem ?

dfilatov replied:
domElem -- это jQuery-chain. http://api.jquery.com/Types/#jQuery.

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+   и передавать в качестве параметров произвольный JS-код.
+
+*NB: на некоторых уровнях переопределения для передачи блока может использоваться другой HTML-атрибут. Например, в 
+библиотеке bem-bl для блоков сенсорных интерфейсов используется атрибут `ondblclick`.*
```
"сенсорных интерфейсов"? это звучит в духе "манипулятор типа мышь"

maslinych replied:
А как нужно написать?

dfilatov replied:
мне кажется, тач-интерфейсы будем всем понятно

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+модификаторами элементов блока. В качестве первого (необязательного)
+параметра все методы принимают имя элемента:
+
+* `{BEM} this.SetMod([elem], modName, modValue)` — Установить модификатор
```
SetMod -> setMod

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+
+* `{BEM} this.SetMod([elem], modName, modValue)` — Установить модификатор
+  `modName` в значение `modValue`. Возвращает JS-объект блока.
+* `{Boolean} this.hasMod([elem], modName, [modValue])` — Проверить
```
Обычно пишем не `modValue`, а `modVal`

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+Новое значение модификатора присваивается блоку *после выполнения всех
+триггеров*. Если необходимо выполнить какие либо операции после
+установки модификатора, следует использовать метод
+`nextTick`. Подробнее см. FIXME.
```
Не-не, `nextTick` для этого делать не надо, это плохой способ. Теперь хуки на установку модификаторов разбиты на две фазы -- `beforeSetMod` (случается до реальной установки модификатора и с помощью `return false` может отменить его установку), `onSetMod` (случается уже после установки, и отменить ничего не может).

maslinych replied:
А конструктор блока теперь нужно писать как beforeSetMod: { 'js': { 'inited': ... } } ? Или оставить onSetMod — если да, то почему? Аналогичный вопрос с деструктором блока.

dfilatov replied:
в onSetMod все-таки правильнее, так как ты пишешь именно обработчик уже случившегося факта. А в beforeSetMod  должно писаться только поведение, которое предназначено именно для возможности отмены установки модификатора.

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+
+# События #
+	
+В i-bem.js поддерживается три(?) вида событий:
```
два, DOM и BEM. live-события могут быть как DOM, так и BEM.

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+	* `{String} event` — Имя события.
+	* `{Object} [data]` — Данные, передаваемые обработчику
+      события. Необязательный параметр. **FIXME**: в каком виде их
+      получает обработчик?
```
в таком же:
`function(e, data) {}`

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+Для привязки обработчиков к BEM-событиям блоков используются методы
+экземпляра блока `this.on` и `this.un`:
+
+* `on([ctx], event, [data], handler, [fnCtx])` — Подписка на
```
параметр `ctx`, может только быть при статических биндах, когда биндишься не к конкретному блоку, а ко всем блокам с заданным именем в пределах контекста `ctx`.

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+и с обычными блоками. В частности, структурировать код в терминах элементов и модификаторов и 
+гибко переопределять поведение библиотеки на разных уровнях переопределения.
+
+Служебным блокам в БЭМ принято давать имена с префиксом `i-`. Таким образом, имя `i-bem.js` 
```
Это вроде уже не верно, i-bem остался единственным блоком с этим префиксом, так типа является именем собственным.

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+        provide(DOM);
+        });
+
+Блоки, не имеющие DOM-представления, декларируются как переопределение ymaps-модуля `i-bem`. 
```
Скорее не как "переопределение", а как "доопределение"

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+декларации блока. Для этого в хэше методов экземпляра блока
+зарезервированы свойства:
+
+* `onSetMod` — триггеры, вызываемые при установке
```
не забыть тут написать про `beforeSetMod` и `beforeElemSetMod`

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1185 @@
+## Доступ к блокам без DOM-представления ##
+
+
+JS-объекты экземпляров блоков, [не привязанных к DOM-дереву](#i-blocks), хранятся в
```
в BEM.blocks хранятся не экземпляры, а "классы" блоков. Экземпляры нигде не хранятся.

maslinych committed:
work through @dfilatov comments on i-bem.js docs

    - rewrite section on modifiers (beforeSetMod)
    - add note on static binding of BEM events
    - smaller fixes in text
    - highlight FIXMEs and add some new ones

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+
+i-bem.js не может:
+
+ * работать с состоянием на сервере (только клиентский JavaScript);
```
// cc @dfilatov 
а точно уместно это писать, если мы говорим, что оторвали jquery из i-bem, чтобы использовать его и на серверной стороне?

dfilatov replied:
да, не очень уместно

maslinych replied:
А если уточнить: i-bem.js не предоставляет инструментария для взаимодействия кода на клиенте с состоянием на сервере ?

tadatuta replied:
в самом ядре не предоставляет, но с использованием его могут быть написаны блоки, которые будут предоставлять. хочется не сбить читателя с толку. может лучше вообще выкинуть эту фразу?

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+i-bem.js не может:
+
+ * работать с состоянием на сервере (только клиентский JavaScript);
+ * заменить высокоуровневые фреймворки общего назначения, подобные jQuery.
```
кажется, что здесь наоборот должно быть «низкоуровневые». т.е. `i-bem.js` — это высокоуровневый БЭМ-ориентированный слой абстракции над низкоуровневым jquery, работающим напрямую с DOM-деревом.

maslinych replied:
уберу этот эпитет совсем, а то любители jQuery могут обидеться

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+ а `button.js` — их логику работы.  
+ + На каждой странице может размещаться более одного **экземпляра** блока. Каждому экземпляру блока соответствует JS-объект, 
+ динамически создаваемый в памяти браузера и хранящий состояние данного экземпляра блока.
+ + **Элементы** — Особые поля в структуре данных экземпляра блока.  
```
// cc @dfilatov Если я правильно понимаю, по умолчанию (без использования `_elem-instances`) для элементов ничего в структуре блока не заводится?

dfilatov replied:
Если отбросить кэш, то да, не заводится. Доступ к элементам осуществляется всегда с помощью методов.

maslinych replied:
А можно поподробнее: чей модификатор `_elem-instances` и что заводится, если его указать?

maslinych replied:
Еще вопрос: могут ли блоки без DOM-представления иметь элементы?

dfilatov replied:
Ну теоретически могут, но i-bem не предоставляет никакой реализации для этого.

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+ + На каждой странице может размещаться более одного **экземпляра** блока. Каждому экземпляру блока соответствует JS-объект, 
+ динамически создаваемый в памяти браузера и хранящий состояние данного экземпляра блока.
+ + **Элементы** — Особые поля в структуре данных экземпляра блока.  
+ + **Модификаторы** — Особые поля экземпляра блока, хранящие информацию о состоянии блока и его элементов.
```
Хотя они и кладутся в кэш, но кажется, что определение модификаторов через поля экземпляра блока не совсем верно.

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+
+Можно использовать i-bem.js как часть полного стека БЭМ-инструментов. Свой проект удобно создавать 
+на основе шаблонного репозитория [project-stub](http://github.com/bem/project-stub/), в котором 
+подключены все необходимые библиотеки и настроена установка БЭМ-инструментов. 
```
БЭМ-инструментов -> [bem-tools](ru.bem.info/tools/bem)?

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+## Сборка ##
+
+Разработка в рамках БЭМ-методологии ведется модульно: каждый блок программируется отдельно, финальный исходный
+код веб-страниц формируется из описаний блоков с помощью процедур **сборки**.
```
описаний блоков -> деклараций бандлов (страниц)?

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+            otherblock.bemhtml
+            ...
+
+Для каждой веб-страницы код использованных на ней блоков должен быть собран в единые файлы: 
```
должен быть -> может быть
(т.к. например, можно еще собирать `merged bandle` общий для всех страниц)

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+Автоматизацию сборки кода результирующих веб-страниц из отдельных описаний блоков осуществляют
+инструменты пакета [bem-tools](http://github.com/bem/bem-tools/). 
+
+*?? Можно ли пользоваться i-bem.js, не используя bem-tools?*
```
можно использовать альтернативный сборщик (например, enb)

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+указанному HTML-фрагменту. 
+
+ * Первичным является DOM-дерево документа, в котором размечены блоки, элементы и модификаторы (в атрибутах `class` HTML-элементов). 
+ * Таким образом, в DOM-дереве каждому экземпляру блока соответствует DOM-элемент.
```
Кажется, что эта фраза может чуть сбивать с толку тем, что есть блоки без DOM-представления, а она звучит, как будто блок всегда имеет DOM-представление.

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+ 2. **Инициализировать экземпляр блока**.<br/> 
+ Включить класс `i-bem` в список классов HTML-элемента. Наличие этого класса позволит фреймворку инициализировать блок.
+ 3. **Передать параметры экземпляру блока**.<br/>
+ Поместить параметры блока в атрибут `onclick`. Параметры блока представляют собой JS-хэш структуры __имя блока—хэш параметров__. 
```
Обязательный вопрос "почему в onclick" задают на каждой конференции. Думаю, сноска с пояснением бы пригодилась.

tadatuta replied:
Нашел описание ниже. Тогда, может стоит поставить ссылку на него здесь?

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+### Один JS-блок на нескольких DOM-элементах ###
+
+Такой дизайн позволяет прозрачно реализовывать блоки, состоящие из нескольких компонент, 
+сосотяние которых должно быть согласовано. Например, виджет «вкладки», где клик по заголовку 
```
по-моему "табы" при всей его сленговости более понятное название

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+ * В HTML-коде страницы не указываются.
+ * JS-код блока (декларация) пишется с использованием несколько иного синтаксиса, чем для блоков с DOM-представлением. Подробнее см. FIXME. 
+
+**NB**: В общем случае инфраструктурный код следует оформлять не в виде блока без DOM-представления, а в виде ymaps-модуля. Например: 
```
imho, фраза предполагает описание того частного случая, когда модуль не подходит и нужен блок без DOM-представления.

maslinych replied:
а что это за частный случай? опишите, пожалуйста, условия, при которых это нужно.

tadatuta replied:
например, в ситуации, когда удобно работать в терминах состояний блока, тогда можно использовать апи i-bem.js для работы с модификаторами

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+Чтобы декларировать новый JS-блок, необходимо переопределить ymaps-модуль `i-bem__dom`. 
+
+Блоки, имеющие DOM-представление (привязанные к DOM-элементу) декларируются с помощью 
+метода `DOM.decl`, принимающим три параметра: 
```
Кажется, что точнее будет сказать:

Чтобы декларировать новый JS-блок с **DOM-представлением (привязанные к DOM-элементу)**, необходимо переопределить ymaps-модуль `i-bem__dom`.
Блоки, декларируются с помощью метода **decl**, принимающего три параметра:

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+        });
+
+Блоки, не имеющие DOM-представления, декларируются как доопределение ymaps-модуля `i-bem`. 
+Вместо метода `DOM.decl` используется метод `BEM.decl`, принимающий те же параметры: 
```
Это ведь не принципиально, т.к. название объекта, в который мы получаем `i-bem` потенциально может быть произвольным.

maslinych replied:
не понял, что именно не принципиально и что нужно изменить в тексте?

tadatuta replied:
Метод называется `decl`, а объект, у которого этот метод вызывается мы можем назвать произвольно, передав его в коллбеке:

``` javascript
 modules.define('i-bem__dom', function(provide, anyNameYouLikeHere) {

        anyNameYouLikeHere.decl(/* селектор блока \*/,
            { 
                /* методы экземпляра \*/
            },
            {
                /* статические методы \*/
            }
        );

        provide(anyNameYouLikeHere);
        });
```

понятно, что там всегда будет `DOM`, но метод decl вызвается у `i-bem` (+ `i-bem__dom`).

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+Селектор представляет собой описание блока, в котором будут применяться
+объявленные в декларации методы.  Селектор обязательно содержит имя блока и 
+может содержать условия, ограничивающие сферу действия методов определенной
+модификацией блока или наличием других блоков на том же DOM-узле.
```
> или наличием других блоков на том же DOM-узле

ничего про не слышал и примера в тексте тоже нет. Такое точно есть?

dfilatov replied:
findBlockOn

maslinych replied:
здесь про наличие блоков на узле — ошибка, исправил

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+**модификаторами**.
+
+* Модификатор это **имя** и **значение**. Например, `size`: `m`.
+* Модификатор может не иметь значения. Например, `disabled`.
```
У такого модификатора неявно предполагается значение `true`.

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+
+	onSetMod: {
+		'js': {
+			'inited': function() { ... } // конструктор экземпляра блока
```
в последней редакции кодстайла (вроде) договорились, что имена ключей не берем в кавычки

maslinych replied:
скажите точно, как договорились, тогда я исправлю по всему тексту

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+
+При автоматической инициализации в памяти браузера будут созданы
+JS-объекты для всех DOM-узлов, в атрибуте `class` которых указан класс
+`i-bem`, а также для всех блоков, задекларированных как блоки, [не
```
Разве для блоков без ДОМ-представления не нужно будет руками звать `create`?

maslinych replied:
Видимо нужно. Можете дать минимальный пример, как инициализация блока без ДОМ-представления выглядит в коде?

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+* `liveInitOnEvent('событие', callback)` — инициализация по событию на блоке или его
+  элементе.<br/>
+  Callback-функция будет вызвана один раз при инициализации блока.
+* `liveBindTo('событие', callback)` — подписка на событие.<br/>
```
Существуют и другие хелперы

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+      передаваемый обработчику в качестве аргумента).
+	* `{Function} handler` — Функция-обработчик события.
+	* `{Object} fnCtx` — Контекст функции-обработчика. Необязательный
+      параметр. **FIXME** зачем его передавать?
```
Чтобы в обработчик в `this` положить нужный объект (или вопрос был не об этом?)

maslinych replied:
Вопрос был о том, в каких ситуациях может быть нужно класть что-то специальное в this?

dfilatov replied:
Когда ты подписываешься на события блока A в блоке B, обычно нужно чтобы контекстом обработчика был блок B, а не блока A.

``` js
DOM.decl('B', {
    method : function() {
        BEM.blocks['A'].on(this.domElem, 'event', this._onEvent, this); // вот тут надо указать этот параметр, чтобы this._onEvent выполнился в контексте экземпляра блока B.
    });
```

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+DOM-узла текущего блока:
+
+* «Внутри блока» — на DOM-узлах, вложенных в DOM-узел текущего блока.
+* «Снаружи блока» — на DOM-узлах, потомком которых является DOM-узел
```
Кажется, тут стоит явно сказать, что желание поискать что-то снаружи скорее всего указывает на архитектурную проблему

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+блоков от других. Однако на практике идеал полной независимости блоков
+недостижим.
+
+Чтобы реализовать взаимодействие блоков, необходимо получить JS-объект
```
Нужно еще сказать, что блоки могут взаимодействовать через события (в т.ч. через каналы), что зачастую более предпочтительно.

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+
+-------------------------------------------------------------------------------
+
+**NB** Не используйте jQuery-селекторы для поиска блоков и элементов.
```
Кажется, что стоит объяснить почему.

maslinych replied:
помогите кратко объяснить

tadatuta replied:
при использовании jquery напрямую нарушается абстракция: мы из предметной области БЭМ вместо обращения к БЭМ-дереву обращаемся к DOM-дереву, повышается связанность кода + потенциально мы когда-нибудь можем начать делать всякие автоматические преобразования кода и хардкод в виде прямых обращений к классам не позволит, скажем, заменить классы на селекторы по кастомным тегам (из `<div class="my-long-class-name"></div>` можно хотеть делать что-то типа `<cn>`) и т.д.

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+      следует вернуть элемент. Необязательный параметр. Подразумевает
+      указание `modName`.
+
+	Возвращает jQuery-коллекцию **FIXME** DOM-элементов(?).
```
да, верно.

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+
+### Параметры блока ###
+
+* getDefaultParams
```
Здесь не хватает метода про получение параметров блока (он в bem-core называется `getParams`?)

dfilatov replied:
параметры находятся в поле `params`

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1208 @@
+
+### Доступ к DOM-узлу блока ###
+
+* containsDomElem
```
Это не про доступ к ДОМ-узлу, а про то, находится ли внутри данного блока определенный ДОМ-элемент.

maslinych committed:
work through @tadatuta comments on i-bem.js docs

tadatuta commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1270 @@
+
+## Методы и свойства модуля BEM ##
+
+**FIXME** нужно ли документировать, стоит
```
считаю, что документировать нужно

maslinych committed:
work through @veged and @dfilatov comments

maslinych committed:
temp commit

maslinych committed:
more changes based on @veged comments

new sections:

    - delegated-events
    - init-ajax
    - init-wave

structural changes:

    - events go before states
    - init section rearranged

assorted smaller fixes

maslinych committed:
rewrite sections on delegated events

    plus minor enhancements in markup and wording

maslinych committed:
i-bem.js reference: bem event object

maslinych committed:
refactor examples

    * in init bem blocks section
    * in find blocks section
    * also some text embellishments

maslinych committed:
i-bem.js reference: clear all fixmes

    + new section: work with DOM
        - elements (moved here)
        - domElem (made a separate section)
        - AJAX (new)
    + remove reference section (to be replaced by autogenerated jsdoc)
    + clear all remaining fixmes

dfilatov commented on common.docs/i-bem.js/i-bem.js.ru.md:
```diff
@@ -0,0 +1,1587 @@
+		'val2': function(modName, modVal, preModVal) { ... }, // триггер на установку mod2 в значение val2
+		'': function(modName, modVal, preModVal) { ... } // триггер на удаление модификатора mod2
+	'mod3': {
+		true: function(modName, modVal, preModVal) { ... }, // триггер на установку простого модификатора mod3
```
К сожалению, из-за IE < 9, придется, все-таки, писать `'true'` вместо `true` (

maslinych committed:
i-bem.js reference: fixes

    * replace "DOM-node" with "HTML-element" where appropriate
    * fix erroneously replaced _onClick mentions
    * fix JSON syntax for data-bem, edit description of parameters format

maslinych committed:
i-bem.js reference: clear final fixmes

  + cleanup details on delegated events
  + add description of unbinding procedures
  + add separate section with trigger examples
  + add final section «What next?»
  + some more minor fixes in text and examples

maslinych committed:
remove links to deleted reference section

    Until API docs generated from JSDoc comments will be available
    online all links to API refrence are deleted or replaced with links
    to source code.

maslinych committed:
i-bem.js reference: last minute fixes

maslinych committed:
i-bem.js add reference title
